### PR TITLE
Interrupt system for stm32f7-discovery

### PR DESCRIPTION
**Pull request by f-bro**
Adds interrupt support for stm32f7-discovery.

The interrupt system features are:

- **Ownership based interrupt management**. The `InterruptTable` owns the nvic register
and thus it is the only one that can access and change the interrupt controller.

 - **Easy to use closure-based ISR registration**. Closures can be registered as interrupt service routine.

- **Free of data races**. Thanks to Rust `Send` and `Sync` concept, the interrupt system is free of data races. Shared mutable access on a variable must be synchronized with a PrimaskMutex, otherwise the compilation fails.
 
- **Scoped IRSs with access to the enviroment**. It is guaranteed that the closure is unregistered at the end of the scope thus it is safe to access the parent stack in the interrupt service routine.

### CODE REVIEW

File: src/interrupts/interrupt_request.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
+//! Interrupt lines of the stm32f7.
+/// Interrupt lines of the stm32f7.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[repr(u8)]
+pub enum InterruptRequest {
+    /// Window Watchdog interrupt 
+ 	Wwdg = 0,
```
Comment (oli-obk) [inline src/interrupts/interrupt_request.rs]: replace all tabs with 4 spaces

File: src/interrupts/mod.rs [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+use core::marker::PhantomData;
+use core::intrinsics::transmute;
+use self::interrupt_request::InterruptRequest;
+
+pub mod interrupt_request;
+pub mod primask_mutex;
+
+unsafe extern "C" fn dispatcher() {
+    let ipsr: u32;
+    asm!("MRS $0, IPSR"
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: add a comment to explain this `asm!` block ("reads the `IPSR` register")

File: src/interrupts/mod.rs [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+//! - **Ownership based interrupt management**. The `InterruptTable` owns the nvic register
+//! and thus it is the only one that can access and change the interrupt controller.
+//!
+//! - **Easy to use closure-based ISR registration**. Closures can be registered as interrupt service routine.
+//!
+//! - **Free of data races**. Thanks to Rust `Send` and `Sync` concept, the interrupt system is free of data races.
+//! Shared mutable access on a variable must be synchronized with a PrimaskMutex,
+//! otherwise the compilation fails.
+//!
+//! - **Scoped IRSs with access to the enviroment**. It is guaranteed that the closure is unregistered at the end of the scope thus it is safe to
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: Punctuation before `thus`. Also add an earlier line break, github requires scrolling ;)

File: src/interrupts/mod.rs [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+}
+
+/// The error type that can occur when handling with interrupts.
+#[derive(Debug)]
+pub enum Error {
+    /// The error type which is returned when an interrupt is registered that is already being used.
+    InterruptAlreadyInUse(InterruptRequest),
+}
+
+/// The `InterruptHandle`is used to access and configure an activ interrupt.
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: nits:

* space before `is`
* activ**e**

File: src/interrupts/mod.rs [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+    ///             let data = interrupt_table.unregister(interrupt_handle);
+    ///             assert!(data.is_none());
+    /// });
+    /// ```
+    pub fn register<F>(&mut self,
+                              irq: InterruptRequest,
+                              priority: Priority,
+                              isr: F)
+                              -> Result<InterruptHandle<()>, Error>
+        where F: FnMut() + 'a + Send
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: couldn't this function's body simply call `register_owned(irq, priority, (), isr)`?

The `()` will get boxed, but that doesn't even allocate.

That would allow `unregister` to return a value unconditionally, instead of returning an `Option`, which should never be `None` except if called on an `InterruptHandle` created by `register` right now.

File: src/interrupts/mod.rs [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+        match Priority::from_u8(res >> 4) {
+            Ok(priority) => priority,
+            Err(PriorityDoesNotExistError(prio_number)) => {
+                unreachable!("Priority {} does not exist", prio_number)
+            }
+        }
+
+    }
+    
+    /// Clears the pendig state of the interrupt corresponding to the `interrupt_handle`.
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: nit: pendi**n**g

File: src/interrupts/mod.rs [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+        let icpr_num = irq as u8 / 32u8;
+        let icpr_bit = irq as u8 % 32u8;
+
+        self.nvic.icpr[icpr_num as usize].update(|r| {
+            let old = r.clrpend();
+            r.set_clrpend(old | 1 << icpr_bit);
+        });
+    }
+
+    /// Sets the pendig state of the interrupt corresponding to the `interrupt_handle`.
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: pendi**n**g

File: src/interrupts/mod.rs [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+        let ispr_num = irq as u8 / 32u8;
+        let ispr_bit = irq as u8 % 32u8;
+
+        self.nvic.ispr[ispr_num as usize].update(|r| {
+            let old = r.setpend();
+            r.set_setpend(old | 1 << ispr_bit);
+        });
+    }
+
+    /// Returns the pendig state of the interrupt corresponding to the `interrupt_handle`.
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: same

File: src/interrupts/primask_mutex.rs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    ///     // Safe, because 'atomic' block
+    ///     data += 5;    
+    /// });
+    /// // Interrupts are enabled again, if interrupts was enabled before the critical section
+    /// ```
+    pub fn lock<F>(&self, critical_section: F) 
+        where F: FnOnce(&mut T)
+    {
+        // PRIMASK = 1 => Prevents the activation of all exceptions with configurable priority
+        let primask = unsafe { ::cortex_m::register::primask::read() } & 1 == 1;
```
Comment (oli-obk) [inline src/interrupts/primask_mutex.rs]: don't we have a race condition here?

Does anything speak against reading the primask after disabling interrupts?

File: src/lib.rs [hunk 10]
```diff
@@ -48,19 +49,26 @@ pub mod random;
...
+                let _ = writeln!(stdout, "\nPANIC in {} at line {}:", file, line);
+                let _ = writeln!(stdout, "    {}", fmt);
+            }
+        });
+
+        loop {}
     });
 
-    loop {}
+    unreachable!("Lock will end in a endless loop!");
```
Comment (oli-obk) [inline src/lib.rs]: you can get rid of this by having `PrimaskMutex::lock` return the value of its closure.

File: src/interrupts/primask_mutex.rs (hunk 9 above)
Comment (f-bro) [inline src/interrupts/primask_mutex.rs]: I don't think we have a data race there.

Imagine we have to activites `A1` and `A2`, where `A2` is an ISR and have a higher priority than `A1`.
1. `A1` reads the primask register and afterwards A1 is preempted.
2. `A2` reads the the primask register, disables interrupts and enters the critical section.
3. `A2` leaves the critical section and enables interrupts again.
3. `A1` enters the critical section.

Since A2 always leaves the critical section (and enable interrupts again) before A1 is conntinued, the Primask register is always in the same state, as it was before A1 was preempted by A2. I think the important thing is, that A2 always returns, before A1 is running again!

Have I overlooked something?

Interrups are disabled by changing the state of the primask register. (Spacial assembler instruction are used to set and reset the primask register, which is one of the core registers).

Comment (oli-obk) [inline src/interrupts/primask_mutex.rs]: Oh right. I was thinking about threads again... With interrupts this is a non-issue.

File: src/interrupts/mod.rs (hunk 5 above)
Comment (f-bro) [inline src/interrupts/mod.rs]: It's a good point!

I'm not sure how the nested closures will effect performance:
```rust
let isr = unsafe {
            transmute::<Box<FnMut()>, Box<FnMut() + 'static + Send>>(Box::new(
            || {
                match self.data[irq as usize] {
                    // Safe, since the correct type is known
                    Some(ptr) => isr(Box::from_raw(transmute::<*mut (), *mut T>(ptr)).as_mut()),
                    None => unreachable!("No data set"),
                }
            }))
};
```

There is also an unneeded table lookup. I'm not sure if the compiler can optimizes that.

I could study the assembler... or should we just go with the potential extra effort but therefore easier code?

Comment (oli-obk) [inline src/interrupts/mod.rs]: Yea, let's go with clear code

**Comment by oli-obk** [comment]
ping @f-bro

**Comment by f-bro** [comment]
blinking_led fails on travis: chicken or the egg.

See: https://github.com/embed-rs/stm32f7-discovery/issues/37

File: src/interrupts/mod.rs [hunk 11]
```diff
@@ -0,0 +1,577 @@
...
+    ///             // Unregister interrupt and get back the ownership to `data`
+    ///             let data = interrupt_table.unregister(interrupt_handle);
+    ///             assert!(data.is_none());
+    /// });
+    /// ```
+    pub fn register<F>(&mut self,
+                              irq: InterruptRequest,
+                              priority: Priority,
+                              mut isr: F)
+                              -> Result<InterruptHandle<()>, Error>
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: too much indentation

File: src/interrupts/mod.rs [hunk 12]
```diff
@@ -0,0 +1,577 @@
...
+///     });
+///    loop{}
+/// }
+/// ```
+///
+/// # Panics
+/// Panics if an interrupt is enabled and is not disabled after use in `code()`
+pub fn scope<'a,F,C,R>(nvic: &'static mut Nvic, default_handler: F, code: C) -> R 
+    where F: FnMut(u8) + 'a,
+        C: FnOnce(&mut InterruptTable<'a>) -> R
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: Reading your thesis I wondered whether we can remove the `PhantomData` from `InterruptTable` and instead write all methods of `InterruptTable` like `fn foo<'a>(&'a mut self, ...)`. I think that would be simpler and convey the same semantics.

File: src/interrupts/mod.rs [hunk 13]
```diff
@@ -0,0 +1,577 @@
...
+///
+///         let _ = interrupt_table.register(Tim7, P1, || {
+///         hprintln!("Interrupt handler for Tim7");
+///     });
+/// });
+/// ```
+pub struct InterruptTable<'a> {
+    _lifetime: PhantomData<&'a ()>,
+    nvic: &'static mut Nvic,
+    data: [Option<* mut ()>; 98],
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: I don't think this needs to be an `Option` anymore, since we always insert a value now.

File: src/touch.rs [hunk 14]
```diff
@@ -32,7 +32,9 @@ pub fn touches(i2c_3: &mut I2C) -> Result<ArrayVec<[Touch; 5]>, i2c::Error> {
                 number_of_touches = 0;
             }
 
-            for &data_reg in FT5336_DATA_REGISTERS.iter().take(number_of_touches.into()) {
+            for &data_reg in FT5336_DATA_REGISTERS
```
Comment (oli-obk) [inline src/touch.rs]: unrelated change

File: src/lcd/mod.rs [hunk 15]
```diff
@@ -39,7 +39,9 @@ pub struct Lcd {
 
 impl Lcd {
     pub fn set_background_color(&mut self, color: Color) {
-        self.controller.bccr.update(|r| r.set_bc(color.to_rgb()));
+        self.controller
```
Comment (oli-obk) [inline src/lcd/mod.rs]: unrelated change

File: src/interrupts/mod.rs [hunk 16]
```diff
@@ -0,0 +1,577 @@
...
+    /// });
+    /// ```
+    pub fn register<F>(&mut self,
+                              irq: InterruptRequest,
+                              priority: Priority,
+                              mut isr: F)
+                              -> Result<InterruptHandle<()>, Error>
+        where F: FnMut() + 'a + Send
+    {
+        self.register_owned(irq, priority, (), move |_| {isr()})
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: nit: `{}` um `isr()` weglassen

File: src/interrupts/mod.rs [hunk 17]
```diff
@@ -0,0 +1,577 @@
...
+        };
+        
+        // transmute::<Box<FnMut()>, Box<FnMut() + 'static + Send>> is safe, because of the drop implementation of InterruptTable ('static is not needed for closure)
+        // and alway only one isr can access the data (Send is not needed for closure)
+        let isr = unsafe {
+            transmute::<Box<FnMut()>, Box<FnMut() + 'static + Send>>(Box::new(
+            || {
+                match self.data[irq as usize] {
+                    // Safe, since the correct type is known
+                    Some(ptr) => isr(Box::from_raw(transmute::<*mut (), *mut T>(ptr)).as_mut()),
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: I think this must be written as `isr(*(ptr as *mut T))`. Otherwise the `Box` is freed after the first time the interrupt is called. You just want a reference here.

File: src/interrupts/mod.rs [hunk 18]
```diff
@@ -0,0 +1,577 @@
...
+    ///             let data = interrupt_table.unregister(interrupt_handle);
+    ///             assert!(data.is_none());
+    /// });
+    /// ```
+    pub fn unregister<T>(&mut self, interrupt_handle: InterruptHandle<T>) -> T {
+        let irq = interrupt_handle.irq;
+        self.disable_interrupt(irq as u8);
+
+        let data = self.data[irq as usize].unwrap();
+        *unsafe {Box::from_raw(transmute::<*mut (), *mut T>(data))}
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: Don't use `transmute` on raw pointers, use `data as *mut T` instead

File: src/interrupts/mod.rs [hunk 19]
```diff
@@ -0,0 +1,577 @@
...
+                                 -> Result<InterruptHandle<T>, Error>
+        where   T: Send,
+                F: FnMut(&mut T) + 'a + Send
+    {
+        if unsafe{ISRS[irq as usize].is_some()} {
+            return Err(Error::InterruptAlreadyInUse(irq));
+        }
+        // Insert data only, when interrupt isn't used, therefore nobody reads the data => no dataraces
+        self.data[irq as usize] = unsafe {
+            transmute::<Option<*mut T>, Option<*mut ()>>(Some(Box::into_raw(Box::new(owned_data))))
```
Comment (oli-obk) [inline src/interrupts/mod.rs]: don't use transmute on raw pointers.

`Some(Box::into_raw(Box::new(owned_data)) as *mut ())` should totally suffice (or without the `Some` after the `Option` removal refactoring)
